opentelemetry-sdk/src/opentelemetry/sdk/_metrics/_internal/metric_reader_storage.py

```diff
@@ -116,13 +128,89 @@ def collect(self) -> Iterable[Metric]:
         # streams produced by the SDK, but we still align the output timestamps
         # for a single instrument.
         with self._lock:
+
+            scope_metrics: List[ScopeMetrics] = []
+
             for (
-                view_instrument_matches
-            ) in self._instrument_view_instrument_matches.values():
+                instrument,
+                view_instrument_matches,
+            ) in self._instrument_view_instrument_matches.items():
+                aggregation_temporality = self._instrument_class_temporality[
+                    instrument.__class__
+                ]
+
+                metrics: List[Metric] = []
+
                 for view_instrument_match in view_instrument_matches:
-                    metrics.extend(view_instrument_match.collect())
 
-        return metrics
+                    if isinstance(
+                        # pylint: disable=protected-access
+                        view_instrument_match._aggregation,
+                        _SumAggregation,
+                    ):
+                        data = Sum(
```

Comment:
The reason to have Sum/Gauge/Histogram pushed all the way down to the Aggregation is to avoid these isinstance checks. The `Aggregation.collect()` method is fully aware of which data type (Sum/Histogram/Gauge) it should make.

Choosing the data point type at this level means we can't support custom aggregators if those get added to the spec, although I guess we can address that if needed.

I think the ViewInstrumentMatch should have all the info it needs to produce the data type. Maybe we could add a classmethod on Aggregation to convert a bunch of "data points" (e.g. NumberDataPoint) to the correct "data" (Sum/Gauge/etc.)

Reply:
I see your point, I agree and I tried fixing it:

There should be a place where we "map" the metric type (`Sum`, `Histogram`, `Gauge`) with the respective aggregation (`_SumAggregation`, `_ExplicitBucketHistogramAggregation`, `_LastValueAggregation`). To be cohesive, we should not have an additional mapping structure (like a dict) or a series of `if isinstance` (as we do right now) but we should do that in the classes, for example by doing this:

``` python
from opentelemetry.sdk._metrics._internal.point import Sum

class _Aggregation(ABC):

    def _get_metric_type(self) -> type:
        pass
        
class _SumAggregation(_Aggregation):

    def _get_metric_type(self) -> type:
        return Sum
```
In this way, instead of doing the series of `if isinstance` that we do now we could simply do this:

```
data = view_instrument_match._aggregation.get_metric_type(...)
```

That would instantiate the right metric. The problem is that the parameters that need to be passed to the constructor of the metric are different depending on the aggregation type itself, for example with a `_SumAggregation` we would have a `Sum` that needs `data_points`, `is_monotonic` and `aggregation_temporality` and with a `_LastValueAggregation` we would have a `Gauge` which only needs `data_points` which again forces us to check the instance type, making all the previous changes irrelevant.

This is a bit of a convoluted problem because we also have to take into consideration `DropAggregation` which should not create a data, again forcing us to check the instance type or doing something else.

I wish this had an elegant solution that would handle all the aggregations in the same way without any type checking, if new aggregations appear in the future we can always just add another check (I know, not pretty) or maybe we can find a better refactor then, but for the time being I am ok with the implementation as it is right now.